File: src/jfan/fan/sys/Depend.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   3 Nov 06  Brian Frank  Creation
//
package fan.sys;

import java.util.*;

/**
 * Depend
 */
public final class Depend
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static Depend fromStr(String str) { return fromStr(str, true); }
  public static Depend fromStr(String str, Boolean checked)
  {
    try
    {
      return new Parser(str).parse();
    }
    catch (Throwable e)
    {
      if (!checked) return null;
      throw ParseErr.make("Depend", str).val;
    }
  }

  private Depend(String name, Constraint[] constraints)
  {
    this.name = name;
    this.constraints = constraints;
  }

//////////////////////////////////////////////////////////////////////////
// Parser
//////////////////////////////////////////////////////////////////////////

  static class Parser
  {
    Parser(String str)
    {
      this.str = str;
      this.len = str.length();
      consume();
    }

    Depend parse()
    {
      name = name();
      constraints.add(constraint());
      while (cur == ',')
      {
        consume();
        consumeSpaces();
        constraints.add(constraint());
      }
      if (pos <= len) throw new RuntimeException();
      return new Depend(name, (Constraint[])constraints .toArray(new Constraint[constraints.size()]));
    }

    private String name()
    {
      StringBuilder s = new StringBuilder();
      while (cur != ' ')
      {
        if (cur < 0) throw new RuntimeException();
        s.append((char)cur);
        consume();
      }
      consumeSpaces();
      if (s.length() == 0) throw new RuntimeException();
      return s.toString();
    }

    private Constraint constraint()
    {
      Constraint c = new Constraint();
      c.version = version();
      consumeSpaces();
      if (cur == '+')
      {
        c.isPlus = true;
        consume();
        consumeSpaces();
      }
      else if (cur == '-')
      {
        consume();
        consumeSpaces();
        c.endVersion = version();
        consumeSpaces();
      }
      return c;
    }

    private Version version()
    {
      List segs = new List(Sys.IntType, 4);
      int seg = consumeDigit();
      while (true)
      {
        if ('0' <= cur && cur <= '9')
        {
          seg = seg*10 + consumeDigit();
        }
        else
        {
          segs.add(Long.valueOf(seg));
          seg = 0;
          if (cur != '.') break;
          else consume();
        }
      }
      return new Version(segs);
    }

    private int consumeDigit()
    {
      if ('0' <= cur && cur <= '9')
      {
        int digit = cur - '0';
        consume();
        return digit;
      }
      throw new RuntimeException();
    }

    private void consumeSpaces()
    {
      while (cur == ' ') consume();
    }

    private void consume()
    {
      if (pos < len)
      {
        cur = str.charAt(pos++);
      }
      else
      {
        cur = -1;
        pos = len+1;
      }
    }

    int cur;
    int pos;
    int len;
    String str;
    String name;
    ArrayList constraints = new ArrayList(4);
  }

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public Boolean _equals(Object obj)
  {
    if (obj instanceof Depend)
      return toStr().equals(toStr(obj));
    else
      return false;
  }

  public int hashCode()
  {
    return toStr().hashCode();
  }

  public Long hash()
  {
    return FanStr.hash(toStr());
  }

  public Type type()
  {
    return Sys.DependType;
  }

  public String toStr()
  {
    if (str == null)
    {
      StringBuilder s = new StringBuilder();
      s.append(name).append(' ');
      for (int i=0; i<constraints.length; ++i)
      {
        if (i > 0) s.append(',');
        Constraint c = constraints[i];
        s.append(c.version);
        if (c.isPlus) s.append('+');
        if (c.endVersion != null) s.append('-').append(c.endVersion);
      }
      str = s.toString();
    }
    return str;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public final String name()
  {
    return name;
  }

  public final Long size()
  {
    return Long.valueOf(constraints.length);
  }

  public final Version version() { return version(0L); }
  public final Version version(Long index)
  {
    return constraints[index.intValue()].version;
  }

  public final Boolean isPlus() { return isPlus(0L); }
  public final Boolean isPlus(Long index)
  {
    return constraints[index.intValue()].isPlus;
  }

  public final Boolean isRange() { return isRange(0L); }
  public final Boolean isRange(Long index)
  {
    return constraints[index.intValue()].endVersion != null;
  }

  public final Version endVersion() { return endVersion(0L); }
  public final Version endVersion(Long index)
  {
    return constraints[index.intValue()].endVersion;
  }

  public final Boolean match(Version v)
  {
    for (int i=0; i<constraints.length; ++i)
    {
      Constraint c = constraints[i];
      if (c.isPlus)
      {
        // versionPlus
        if (c.version.compare(v).longValue() <= 0)
          return true;
      }
      else if (c.endVersion != null)
      {
        // versionRange
        if (c.version.compare(v).longValue() <= 0 &&
            (c.endVersion.compare(v).longValue() >= 0 || doMatch(c.endVersion, v)))
          return true;
      }
      else
      {
        // versionSimple
        if (doMatch(c.version, v))
          return true;
      }
    }
    return false;
  }

  private static boolean doMatch(Version a, Version b)
  {
    if (a.segments().sz() > b.segments().sz()) return false;
    for (int i=0; i<a.segments().sz(); ++i)
      if (a.segment(i) != b.segment(i))
        return false;
    return true;
  }

//////////////////////////////////////////////////////////////////////////
// Constraint
//////////////////////////////////////////////////////////////////////////

  static class Constraint
  {
    Version version;
    boolean isPlus;
    Version endVersion;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  private final String name;
  private final Constraint[] constraints;
  private String str;

}

File: src/sys/es/fan/Depend.js
//
// Copyright (c) 2011 Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   03 Jan 2011  Andy Frank  Creation
//   20 Apr 2023  Matthew Giannini  Refactor for ES
//

/**
 * Depend.
 */
class Depend extends Obj {

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  constructor(name, constraints) {
    super();
    this.#name = name;
    this.#constraints = constraints;
    this.#str = null;
  }

  #name;
  #constraints;
  #str;

  static fromStr(str, checked=true) {
    try {
      // allow try-block to capture errs
      return new DependParser(str).parse();
    }
    catch (err) {
      if (!checked) return null;
      throw ParseErr.makeStr("Depend", str);
    }
  }

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  equals(obj) {
    if (obj instanceof Depend)
      return this.toStr() == obj.toStr();
    else
      return false;
  }

  hash() {
    return Str.hash(this.toStr());
  }

  toStr() {
    if (this.#str == null) {
      let s = "";
      s += this.#name + " ";
      for (let i=0; i<this.#constraints.length; ++i) {
        if (i > 0) s += ",";
        var c = this.#constraints[i];
        s += c.version;
        if (c.isPlus) s += "+";
        if (c.endVersion != null) s += "-" + c.endVersion;
      }
      this.#str = s.toString();
    }
    return this.#str;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  name() { return this.#name; }

  size() { return this.#constraints.length; }

  version(index=0) { return this.#constraints[index].version; }

  isSimple(index=0) { return !this.isPlus(index) && !this.isRange(index); }

  isPlus(index=0) { return this.#constraints[index].isPlus; }

  isRange(index=0) { return this.#constraints[index].endVersion != null; }

  endVersion(index=0) { return this.#constraints[index].endVersion; }

  match(v) {
    for (let i=0; i<this.#constraints.length; ++i) {
      const c = this.#constraints[i];
      if (c.isPlus) {
        // versionPlus
        if (c.version.compare(v) <= 0)
          return true;
      }
      else if (c.endVersion != null) {
        // versionRange
        if (c.version.compare(v) <= 0 &&
            (c.endVersion.compare(v) >= 0 || Depend.#doMatch(c.endVersion, v)))
          return true;
      }
      else
      {
        // versionSimple
        if (Depend.#doMatch(c.version, v))
          return true;
      }
    }
    return false;
  }

  static #doMatch(a, b) {
    if (a.segments().size() > b.segments().size()) return false;
    for (let i=0; i<a.segments().size(); ++i)
      if (a.segment(i) != b.segment(i))
        return false;
    return true;
  }

}

//////////////////////////////////////////////////////////////////////////
// DependConstraint
//////////////////////////////////////////////////////////////////////////

class DependConstraint {
  constructor() {
    this.version = null;
    this.isPlus = false;
    this.endVersion = null;
  }

  version;
  isPlus;
  endVersion;
}

//////////////////////////////////////////////////////////////////////////
// DependParser
//////////////////////////////////////////////////////////////////////////

class DependParser {
  constructor(str) {
    this.str = str;
    this.cur = 0;
    this.pos = 0;
    this.len = str.length;
    this.constraints = [];
    this.consume();
  }

  str;
  cur;
  pos;
  len;
  constraints;

  parse() {
    const name = this.#name();
    this.constraints.push(this.constraint());
    while (this.cur == 44) {
      this.consume();
      this.consumeSpaces();
      this.constraints.push(this.constraint());
    }
    if (this.pos <= this.len) throw new Error();
    return new Depend(name, this.constraints);
  }

  #name() {
    let s = ""
    while (this.cur != 32) {
      if (this.cur < 0) throw new Error();
      s += String.fromCharCode(this.cur);
      this.consume();
    }
    this.consumeSpaces();
    if (s.length == 0) throw new Error();
    return s;
  }

  constraint() {
    let c = new DependConstraint();
    c.version = this.version();
    this.consumeSpaces();
    if (this.cur == 43) {
      c.isPlus = true;
      this.consume();
      this.consumeSpaces();
    }
    else if (this.cur == 45) {
      this.consume();
      this.consumeSpaces();
      c.endVersion = this.version();
      this.consumeSpaces();
    }
    return c;
  }

  version() {
    const segs = List.make(Int.type$);
    let seg = this.consumeDigit();
    while (true) {
      if (48 <= this.cur && this.cur <= 57) {
        seg = seg*10 + this.consumeDigit();
      }
      else {
        segs.add(seg);
        seg = 0;
        if (this.cur != 46) break;
        else this.consume();
      }
    }
    return Version.make(segs);
  }

  consumeDigit() {
    if (48 <= this.cur && this.cur <= 57) {
      const digit = this.cur - 48;
      this.consume();
      return digit;
    }
    throw new Error();
  }

  consumeSpaces() {
    while (this.cur == 32 || this.cur == 9) this.consume();
  }

  consume() {
    if (this.pos < this.len) {
      this.cur = this.str.charCodeAt(this.pos++);
    }
    else {
      this.cur = -1;
      this.pos = this.len+1;
    }
  }
}

